src/dataprep/main/prep_nsf/link_scinetnsf_to_mag.R

```diff
@@ -0,0 +1,195 @@
+  first_row_similarity <- stringsim(
+    mag_firstname,
+    nsf_firstname,
+    method="osa" )
+  
+  
+  mag_lastname <- row$mag_lastname
+  nsf_lastname <- row$nsf_lastname
+  
+  last_row_similarity <- stringsim(
+    mag_lastname,
+    nsf_lastname,
+    method = "osa"
+  )
+  
+    return(data.frame(firstname_similarity = first_row_similarity, lastname_similarity = last_row_similarity))
+}
+
+# Split the data into chunks of 50,000 rows
+chunk_size <- 50000
+chunks <- split(nsf_to_authors, ceiling(seq_len(nrow(nsf_to_authors)) / chunk_size))
+
+# Load the furrr package for parallel processing
+plan(multisession, workers=16)
+
+# Initialize variables for progress tracking
+total_chunks <- length(chunks)
+processed_chunks <- 0
+
+# Process and save each chunk as individual CSV files
+ time <- Sys.time()
+ cat(sprintf(
+  "Start processing %d chunks at %s \n", total_chunks,	start_time))
+
+for (i in seq_along(chunks)) {
+  chunk <- chunks[[i]]
+  
+  # Calculate similarity and filter rows row by row
+  row_similarities <- purrr::map_df(1:nrow(chunk), ~fct_similarity(chunk[.x, ])) %>%
+    mutate(id = row_number())
+  
+  # Filter rows that meet the threshold criteria
+  chunk <- chunk %>%
+    mutate(id = row_number()) %>%
+    left_join(row_similarities, by = "id") %>%
+    filter(
+      (firstname_similarity >= threshold & lastname_similarity >= threshold) | (lastname_similarity == 1.0 & substr(mag_firstname, 1, 1) == substr(nsf_firstname, 1, 1)) ) %>%
+    select(GrantID, AuthorId, Position, mag_firstname, nsf_firstname, firstname_similarity, mag_lastname, nsf_lastname, lastname_similarity) %>%
+    distinct()
+    
+  
+  # Define the output file path
+  output_file <- file.path("/mnt/ssd/chunks_nsf_links", paste0("chunk_", i, ".csv"))
+
+# Remove the output file if it exists
+if (file.exists(output_file)) {
+  file.remove(output_file)
+}
+  
+  # Write the chunk to a CSV file
+  write.csv(chunk, file = output_file, row.names = FALSE)
+  
+  # Update progress
+  processed_chunks <- processed_chunks + 1
+  percent_processed <- (processed_chunks / total_chunks) * 100
+  elapsed_time <- as.numeric(Sys.time() - start_time)
+
+
+  # Display progress information
+    cat(sprintf(
+      "Processed %d out of %d chunks (%.2f%%) in %.2f.\n",
+      processed_chunks, total_chunks, percent_processed, elapsed_time
+    ))
+}
+
+# Clean up the furrr plan
+plan(NULL)
+
+# Some info
+final_elapsed_time <- Sys.time() - start_time
+final_elapsed_time <- as.numeric(final_elapsed_time)
+
+# Convert elapsed time to minutes and potentially hours
+
+  # Display progress information
+  cat(sprintf(
+    "Complete. Total elapsed time: %.2f.\n",
+    final_elapsed_time
+  ))
+
+
+# close connection to db
+DBI::dbDisconnect(con)
+cat("Disconnected from db.\n")
+
+# Apend tables together
+# Initialize an empty data frame to store the appended data
+links_nsf_mag <- data.frame()
+
+# Loop through the file names and append the data
+for (i in 1:1072) {
```

Comment:
hard coded loop length. Please replace by something that works generically

Reply:
actually the whole loop etc seems to be a leftover of code from before. Line 176 to end should be deleted, right?